lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2018,6 +2016,47 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
 // Statements
 //===----------------------------------------------------------------------===//
 
+/// If the connect is for RefType, implement the constraint for downward only
+/// references. We cannot connect :
+///   1. an input reference port to the output reference port.
+///   2. instance reference results to each other.
+/// This means, the connect can only be used for forwarding RefType module
+/// ports to Instance ports.
+static LogicalResult checkRefTypeFlow(Operation *connect) {
+  Value dst = connect->getOperand(0);
+  Value src = connect->getOperand(1);
+
+  if (dst.getType().isa<RefType>()) {
+    if (getDeclarationKind(src) == getDeclarationKind(dst)) {
+      auto srcRef = getFieldRefFromValue(src);
+      bool rootKnown;
+      auto srcName = getFieldName(srcRef, rootKnown);
+      auto diag = emitError(connect->getLoc())
+                  << "connect has invalid flow for ref type ports: the source "
+                     "expression ";
+      if (rootKnown)
+        diag << "\"" << srcName << "\" ";
+      diag << "and destination expression ";
+      auto dstRef = getFieldRefFromValue(dst);
+      bool dstRootKnown;
+      auto dstName = getFieldName(dstRef, dstRootKnown);
+      if (dstRootKnown)
+        diag << "\"" << dstName << "\" ";
+      diag << "both have same port kind, expected Module port to Instance "
+              "connections only";
+      return diag;
+    }
+    if ((src.getType().isa<RefType>() && !src.hasOneUse()) ||
+        (dst.getType().isa<RefType>() && !dst.hasOneUse())) {
```

Comment:
We have to remove the restriction on multiple readers for a single XMR ref. But lets remove the restriction on a separate PR. It depends on the XMR Ops. For now, this constraint ensures that none of the Ref ports are reused.
Also maybe, the `RefType` checks can be avoided if we filter `BaseType` at function entry.